src/commands.cc

```diff
@@ -1415,11 +1421,13 @@ auto map_key_completer =
         if (token_to_complete == 0)
             return { 0_byte, params[0].length(),
                      complete(params[0], pos_in_token, scopes) };
+
         if (token_to_complete == 1)
         {
-            constexpr const char* modes[] = { "normal", "insert", "menu", "prompt", "goto", "view", "user", "object" };
+            auto modes = { "normal", "insert", "menu", "prompt", "goto", "view", "user", "object" };
+            auto user_modes = get_scope(params[0], context).keymaps().user_modes();
             return { 0_byte, params[1].length(),
-                     complete(params[1], pos_in_token, modes) };
+                     complete(params[1], pos_in_token, concatenated(modes, user_modes) | gather<Vector<String>>()) };
```

Comment:
`complete` takes an arbitrary container as last argument, so no need to gather the concatenated result into a vector, we can avoid copying all that data.

Reply:
I had to use `gather` because without it the code triggers this static_assert error about long live string:

```
completion.hh: In instantiation of ‘Kakoune::CandidateList Kakoune::complete(Kakoune::StringView, Kakoune::ByteCount, const Container&) [with Container = Kakoune::ConcatView<std::initializer_list<const char*>&, std::vector<Kakoune::String, Kakoune::Allocator<Kakoune::String, (Kakoune::MemoryDomain)0> >&>; Kakoune::CandidateList = std::vector<Kakoune::String, Kakoune::Allocator<Kakoune::String, (Kakoune::MemoryDomain)22> >]’:

commands.cc:1430:87:   required from here

completion.hh:67:5: error: static assertion failed: complete require long lived strings, not temporaries
     static_assert(not std::is_same<decltype(*begin(container)), String>::value,
```

What would be the alternative?

Reply:
Ahh, right, ok I guess you can gather there for now.